lib/spack/spack/util/gpg.py

```diff
@@ -2,74 +2,123 @@
 # Spack Project Developers. See the top-level COPYRIGHT file for details.
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
-
 import contextlib
 import errno
 import functools
 import os
 import re
 
-import llnl.util.lang
-
 import spack.error
 import spack.paths
 import spack.util.executable
 import spack.version
 
 
-_gnupg_version_re = r"^gpg(conf)? \(GnuPG\) (.*)$"
-_gnupg_home_override = None
-_global_gpg_instance = None
+#: Executable instance for "gpg", initialized lazily
+GPG = None
+#: Executable instance for "gpgconf", initialized lazily
+GPGCONF = None
+#: Socket directory required if a non default home directory is used
+SOCKET_DIR = None
+#: GNUPGHOME environment variable in the context of this Python module
+GNUPGHOME = None
 
 
-def get_gnupg_home(gnupg_home=None):
-    """Returns the directory that should be used as the GNUPGHOME environment
-    variable when calling gpg.
+def clear():
+    """Reset the global state to uninitialized."""
+    global GPG, GPGCONF, SOCKET_DIR, GNUPGHOME
+    GPG, GPGCONF, SOCKET_DIR, GNUPGHOME = None, None, None, None
 
-    If a [gnupg_home] is passed directly (and not None), that value will be
-    used.
 
-    Otherwise, if there is an override set (and it is not None), then that
-    value will be used.
+def init(gnupghome=None, force=False):
+    """Initialize the global objects in the module, if not set.
 
-    Otherwise, if the environment variable "SPACK_GNUPGHOME" is set, then that
-    value will be used.
+    When calling any gpg executable, the GNUPGHOME environment
+    variable is set to:
 
-    Otherwise, the default gpg path for Spack will be used.
+    1. The value of the `gnupghome` argument, if not None
+    2. The value of the "SPACK_GNUPGHOME" environment variable, if set
+    3. The default gpg path for Spack otherwise
 
-    See also: gnupg_home_override()
+    Args:
+        gnupghome (str): value to be used for GNUPGHOME when calling
+            GnuPG executables
+        force (bool): if True forces the re-initialization even if the
+            global objects are set already
     """
-    return (gnupg_home or
-            _gnupg_home_override or
-            os.getenv('SPACK_GNUPGHOME') or
-            spack.paths.gpg_path)
+    global GPG, GPGCONF, SOCKET_DIR, GNUPGHOME
+    if force:
+        clear()
 
+    # If the executables are already set, there's nothing to do
+    already_set = GPG and GPGCONF
+    if already_set:
+        return
 
-@contextlib.contextmanager
-def gnupg_home_override(new_gnupg_home):
-    global _gnupg_home_override
-    global _global_gpg_instance
+    # Set the value of GNUPGHOME to be used in this module
+    GNUPGHOME = (gnupghome or
+                 os.getenv('SPACK_GNUPGHOME') or
+                 spack.paths.gpg_path)
 
-    old_gnupg_home_override = _gnupg_home_override
-    old_global_gpg_instance = _global_gpg_instance
+    # Set the executable objects for "gpg" and "gpgconf"
+    GPG, GPGCONF = _gpg(), _gpgconf()
```

Comment:
For reviewers. I started this refactoring to have a single place to modify (the line above) when adding bootstrapping for GnuPG. While I was at it I figured out I could simplify the implementation in this module without changing the API exposed to clients, so I took a stab at it.